### PR TITLE
feat(cicd): run functional on rdev with auth

### PR DESCRIPTION
**Pull request by Bento007**
## Reason for Change

- #5574  
- #5436
- Support running functional tests in an rdev environment.

## Changes
- Use Authentication header to get through oauth_proxy.
- Use STACK_NAME to identify the rdev environment in functional tests
- Support rdev environments for functional tests.
- use a tenacity timeout strategy for failing api key test.
- pass needed environment variables to docker image during test
- use a config dummy file for CorporaAuthConfig to avoid using AWS secret manager for the functional tests. Chamber already handles passing in all the needed secrets.

## Testing steps
- test against existing environment
   ```export STACK_NAME=auth DEPLOYMENT_STAGE=rdev AWS_PROFILE=single-cell-dev && make local-functional-test```
- [testing instructions](https://docs.google.com/document/d/1nynGcBS_TA55qlQo9WjINGkcMnE_xIBz7-inmop2bqo/edit#heading=h.t5qd7armv63z)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -211,13 +211,18 @@ local-unit-test-cxg-admin:
...
-local-functional-test: export AWS_PROFILE=$(TEST_AWS_PROFILE)
 local-functional-test: ## Run functional tests in the dev environment
 	if [ -n "$${BOTO_ENDPOINT_URL+set}" ]; then \
 		EXTRA_ARGS="-e BOTO_ENDPOINT_URL"; \
 	fi; \
 	chamber -b secretsmanager exec corpora/backend/$${DEPLOYMENT_STAGE}/auth0-secret -- \
-		docker-compose $(COMPOSE_OPTS) run --rm -T -e CLIENT_ID -e CLIENT_SECRET -e TEST_ACCOUNT_USERNAME -e TEST_ACCOUNT_PASSWORD -e DEPLOYMENT_STAGE -e AWS_ACCESS_KEY_ID -e AWS_SECRET_ACCESS_KEY -e AWS_SESSION_TOKEN $${EXTRA_ARGS} \
+		docker-compose $(COMPOSE_OPTS) run --rm -T \
+		-e CLIENT_ID -e CLIENT_SECRET \
+		-e FUNCTEST_ACCOUNT_USERNAME -e FUNCTEST_ACCOUNT_PASSWORD \
```
Comment (danieljhegeman) [inline Makefile:221]: this would be for actually logging the test user in (the normal cellxgene auth flow)...

File: Makefile [hunk 2]
```diff
@@ -211,13 +211,18 @@ local-unit-test-cxg-admin:
...
+		-e TEST_AUTH0_USER_ACCOUNT_PASSWORD -e TEST_APP_ID -e TEST_APP_SECRET \
```
Comment (danieljhegeman) [inline Makefile:222]: ...while this would be for getting past the outer auth0 layer / proxy, right?

Comment (danieljhegeman) [reply Makefile:222]: My resulting question then would be: why wasn't the functest_account_username secret being used before now...

File: Makefile (hunk 1 above)
Comment (Bento007) [reply Makefile:221]: yes using cookies

File: Makefile (hunk 2 above)
Comment (Bento007) [reply Makefile:222]: > ...while this would be for getting past the outer auth0 layer / proxy, right?

yes for `-e TEST_APP_ID -e TEST_APP_SECRET`

Comment (Bento007) [reply Makefile:222]: > y resulting question then would be: why wasn't the functest_account_username secret being used before now...

So they actually were being used, but they were getting introduced into the container through a different path. In the [GHA functional test](https://github.com/chanzuckerberg/single-cell-data-portal/blob/main/.github/workflows/deploy-happy-stack.yml.), the container was getting the secret directly from AWS Secrets manager using credentials supplied to the docker container. The AWS credentials were passed into the GHA container here https://github.com/chanzuckerberg/single-cell-data-portal/blob/3bca29bd6cef0b8d0eab6fdd49edf57cd8741364/Makefile#L220
but this didn't work when running the tests locally or when `DEPLOYMENT_STAGE=test` which is why this comment
https://github.com/chanzuckerberg/single-cell-data-portal/blob/3bca29bd6cef0b8d0eab6fdd49edf57cd8741364/Makefile#L206-L212

Now all of the secrets needed by the functional tests are introduced via the environment using the make recipe and chamber. No more sneaking secret getting introduced from multiple different sources. The functional tests no longer need to access AWS secrets to work.

Comment (Bento007) [reply Makefile:214]: I'm going to remove the AWS credentials and the comment from the make recipe. They aren't needed.

Comment (danieljhegeman) [reply Makefile:214]: excellent, thanks!

**Review by danieljhegeman — APPROVED** [review]
worked for me locally 👍

File: tests/functional/backend/common.py [hunk 3]
```diff
@@ -15,22 +17,34 @@
...
         super().setUpClass()
         cls.deployment_stage = os.environ["DEPLOYMENT_STAGE"]
-        cls.config = CorporaAuthConfig()
+
+        # configure CorporaAuthConfig to use a temporary directory for the config file
+        cls.tempdir = tempfile.TemporaryDirectory()
+        dummy_config = f"{cls.tempdir.name}/dummy.json"
+        with open(dummy_config, "w") as fp:
+            json.dump({"api_base_url": os.getenv("API_BASE_URL")}, fp)
+        cls.config = CorporaAuthConfig(source=dummy_config)
```
Comment (seve) [inline tests/functional/backend/common.py:47]: stinky

**Pull request merged**